### ISSUE TITLE
Add some error checking and reporting for common TriBITS project developer and user mistakes

### ISSUE THREAD

**Issue by bartlettroscoe** [TriBITSPub/TriBITS#200]
**Next Action Status:**

Added checks for usage of TRIBITS_PACKAGE(), TRIBITS_PACKAGE_POSTPROCESS(), etc. and some other things.  Next: Need to add checks for TRIBITS_SUBPACKAGE(), TRIBITS_SUBPACKAGE_POSTPROCESS(), etc.

**CC:** @fryeguy52 

**Description:**

The TriBITS framework can reduce some of the boiler plate raw CMake code that project developers need to write when setting up a CMake build and tests with CTest in their CMakeLists.txt files.  However, history has shown that project developers and TriBITS project users made some common mistakes when using TriBITS.  TriBITS already contains error checks and reporting for many types of common mistakes (e.g. misspelling the name of a TriBITS SE package, or trying to create a circular dependency) but there are many other types of mistakes that are not currently caught and reported very gracefully (e.g. not calling TRIBITS_PROJECT_POSTPROCESS() at the end of a packages's CMakeLists.txt file)

This Story is to add error checking and reporting (using `MESSAGE(FATAL_ERROR ...)`) for many common mistakes.  An initial list of the mistakes that can be caught and gracefully reported are:

* Check that `TRIBITS_PACKAGE()` is called before `TRIBITS_ADD_LIBRARY()` or `TRIBITS_ADD_EXECUTABLE()`. **[DONE]**

* Check that `TRIBITS_PACKAGE_DECL()`, `TRIBITS_PACKAGE_DEF()`, `TRIBITS_PACKAGE_POSTPROCESS()`, `TRIBITS_PROJECT()` and many other such user-facing macros and functions are called only once (i.e. per package for package-level functions and once per project for project-level functions). **[DONE]**

* Check that `TRIBITS_PACKAGE_POSTPROCESS()` is called before the end of the package's CMakeLists.txt file after all other commands like `TRIBITS_ADD_LIBRARY()` and `TRIBITS_ADD_EXECUTABLE()`. **[DONE]**

* Check that all of the arguments to `TRIBITS_PROJECT_DEFINE_DEPENDENCIES()` match named arguments (e.g. look for non-empty `PARSE_UNPARSED_ARGUMENTS` after calling `CMAKE_PARSE_ARGUMENTS(PREFIX ...)` and abort if found). **[DONE]**

* Add checks for all TriBITS functions and macros that use `CMAKE_PARSE_ARGUMENTS(PREFIX ...)` for `PARSE_UNPARSED_ARGUMENTS` when they should not exist. **[DONE]**

* Add checks that `TRIBITS_SUBPACKAGE()`, `TRIBITS_SUBPACKAGE_POSTPROCESS()`, etc. are called correctly **[DONE]**

* Check that TRIBITS_PACKAGE() is not called in the base CMakeLists.txt file for a package that has subpackages.  Instead, they must call TRIBITS_PACKAGE_DECL() and TRIBITS_PACKAGE_DEF() (with TRIBITS_PROCESS_SUBPACKAGES() called in between). **[DONE]**

* Check that missing a call to TRIBITS_PROCESS_SUBPACKAGES() is caught if one forgets to call it before TRIBITS_PACKAGE_DEF() is called. **[DONE]**

* Check that if only TRIBITS_PACKAGE_DECL() is called in parent packages's CMakeList.txt file that TriBITS reports the user needs to call TRIBITS_PROCESS_SUBPACKAGES() next. **[DONE]**

* Check that TRIBITS_PACKAGE_DEF() is called once and only once in a parent packages's CMakeLists.txt file. **[DONE]**

* Check that a subpackages's CMakeLists.txt files do not call any of the commands TRIBITS_PACKAGE(), TRIBITS_PACKAGE_DECL(), TRIBITS_PACKAGE_DEF(), TRIBIITS_PROCESS_SUBPACKAGES(), TRIBITS_PACKAGE_POSTPROCESS(). **[DONE]**

* Check that TRIBITS_ADD_TEST_DIRECTORIES() and TRIBITS_ADD_EXAMPLE_DIRECTORIES() are called between TRIBITS_SUBPROJECT() and TRIBITS_SUBPROJECT_POSTPROCESS() (and the same for a top-level packages). **[DONE]**

The above are just a few of the types of mistakes that TriBITS developers and users have made in the past and TriBITS does not catch and report very well. Some more thought and exploration will yield other missing checks.  But this story does not have to involve finding and resolving all such errors.

What makes this story non-trivial is that we must set up automated testing to check that TriBITS can catch and report these mistakes.  To do that, one could set up new testing dummy TriBITS package(s) that can be configured to make all of these mistakes and then add automated tests in the TriBITS test suite to make sure that TriBITS code catches the mistakes and prints the right error messages.

**Comment by bartlettroscoe** [comment]
@fryeguy52, I think this is a natural story to go along with setting up a tutorial for TriBITS in #190.

**Comment by bartlettroscoe** [comment]
@fryeguy52 and I pair programmed the starter to this in:

* https://github.com/TriBITSPub/TriBITS/tree/user-error-checking-200

@fryeguy52, please checkout out this branch in your local git repo on my machine crf450 and continue one.  Let me know if you have questions.

**Comment by bartlettroscoe** [comment]
I just merged the branch 'user-error-checking-200' that takes care of several of these checks (see above ['description' field](https://github.com/TriBITSPub/TriBITS/issues/200#issue-240753985)).

@fryeguy52,

Note that there are still several types of checks mentioned [above](https://github.com/TriBITSPub/TriBITS/issues/200#issue-240753985).

**Comment by bartlettroscoe** [comment]
@fryguy52,

I cherry-picked, tested, and pushed the commit 9c09636 which was pushed as the commit 3559d6b07bd08effd774fdf4c6f3e5f0d86775d0.

I then deleted the user-error-checking-200 branch:

```
$ git push github --delete user-error-checking-200 
To git@github.com:TriBITSPub/TriBITS.git
 - [deleted]         user-error-checking-200
```

For any future work on this story, we should create follow-on topic branches like `assert-single-arg-values-200` to add checking for single-value arguments.

**Comment by bartlettroscoe** [comment]
When I tried to build Trilinos with the new TRIBITS_CHECK_FOR_UNPARSED_ARGUMENTS() macro, I found a lot of violations.  It was very hard to fix because of the FATAL_ERROR as each one was hit and fixed.  Therefore, I added the commit https://github.com/TriBITSPub/TriBITS/commit/822a64a92646e20ab8162d6165fbf073946e5845 which makes this a warning by default but allows you to make it SEND_ERROR or FATAL_ERROR.  I then made the TriBITS project set FATAL_ERROR in commit https://github.com/TriBITSPub/TriBITS/commit/8e1bd0c35b2ee13674521b8063c8438e7d262cb2.

Also, I discovered there are other types of errors that people are making that we are not catching yet. In commit https://github.com/trilinos/Trilinos/commit/1fc79d54b872ba382062a709395e7a9638e89687 I saw an example where TRIBITS_PACKAGE_DECL() was called multiple times.  Therefore, we need to add checking that the commands TRIBITS_PACKAGE_DECL(), TRIBITS_PACKAGE_DEF() and TRIBITS_PACKAGE_POSTPROCESS() are called exactly one time for each package.

**Comment by bartlettroscoe** [comment]
@fryeguy52, we need to add checks for proper usage of TRIBITS_SUBPACKAGE() and TRIBITS_SUBPACKAGE_POSTPROCESS().  A missing TRIBITS_SUBPACKAGE_POSTPROCESS() resulted in a lot of lost time and productivity recently as described in:

* https://github.com/trilinos/Trilinos/issues/1788#issuecomment-334881474

I think that means we need a dummy test packages that has subpackages and then do the same type of testing.  We can copy and paste the TribitsExampleProect WithSubpackages package and then strip it down and dress it up to test and catch these types of errors like we did for TRIBITS_PACKAGE(), TRIBITS_PACKAGE_POSTPROCESS(), etc.

**Comment by bartlettroscoe** [comment]
I just pushed the branch:

* https://github.com/TriBITSPub/TriBITS/tree/subpkg-user-errors-200

```
To git@github.com:TriBITSPub/TriBITS.git
 * [new branch]      subpkg-user-errors-200 -> subpkg-user-errors-200
```

that has a new testing package PkgWithSubpkgsWithUserErrors that we use to add user mistakes with Subpackages and then add checks to catch them, like we did with PkgWithUserErrors for Package errors.

@fryeguy52,

Can you please checkout this branch with:

```
$ cd TriBITS
$ git fetch origin   # Or whatever remote you have for git@github.com:TriBITSPub/TriBITS.git
$ git checkout --track origin/subpkg-user-errors-200
```

and then precede to break it and add checks to gracefully report the failures like you did for PkgWtihUserErrors?

**Comment by bartlettroscoe** [comment]
The subpackage checks are really needed.  Here is another example trilinos/Trilinos#218.  I don't know how code like that even works.

**Comment by bartlettroscoe** [comment]
FYI: A bunch more of the package a subpackage function checks were added as part of #232 (thanks @fryeguy52).

**Comment by bartlettroscoe** [comment]
@fryeguy52, I think we need to improve the error message when someone accidentally calls `TRIBITS_PACAKGE_(<packageName>)` instead of `TRIBITS_PACKAGE_DECL(<packageName>)` in a parent package's CMakeListst.txt file.  The current checks give a confusing error message as reported at:

* https://github.com/SCOREC/core/issues/128

What we need is a clear error message that says "Can't call TRIBITS_PACKAGE() since this parent package has subpackages.  Instead, must call TRIBITS_PACKAGE_DECL(), then TRIBITS_PROCESS_SUBPACKAGES() then TRIBITS_PACKAGE_DEF()".  Can you look into improving that error message?

**Comment by fryeguy52** [comment]
@bartlettroscoe it looks like your commit fixed this. Do you need me to do anything?

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe it looks like your commit fixed this. Do you need me to do anything?

I think I fixed this use case.  If we see other cases like this come up, we may need to dig in and try to improve the error checking and error messages.  It turns out that in the case of SCOREC that it used to have subpackages then it was refactored to not have subpackages but the top-level SCOREC CMakeLists.txt file never got updated.  It turns out that worked just fine.  But the new checking did not allow that use case.  I added a check to disallow that.  But I still allowed one to call TRIBITS_PACKAGE_DECL() and TRIBIT_PACAKGE_DEF() instead of TRIBITS_PACKAGE() for packages with no subpackages.  I am not sure if we should allow that or not but it seems okay to do that.
